feat: Add remote image caching behaviour docs

oliverlynch opened the pull request:
<!-- Thank you for opening a PR! We really appreciate you taking the time to help out 🙌 -->

#### Description (required)

<!-- Please describe the change you are proposing, and why -->

<!-- Please make changes in **one language** only -->

Currently the behaviour of Astro's remote image cache is undocumented, which makes discovering the feature difficult and its behaviour unclear. This change adds an explanation of the asset cache, as well as some technical details on how it functions, which should be helpful for anyone configuring a remote image server for Astro. 

Part of this change references the revalidation feature introduced in Astro PR [#12426](https://github.com/withastro/astro/pull/12426), however the rest documents existing functionality. 

#### Related issues & labels (optional)

- Closes #<!-- Add an issue number if this PR will close it. -->
- Suggested label: add new content <!-- Help us triage by suggesting one of our labels that describes your PR -->

<!-- For a new/changed feature in an upcoming Astro release? -->
<!-- 1. Uncomment the line below, update the minor version number if known, and include a PR link -->
#### For Astro version: `5.1`. See astro PR [#12426](https://github.com/withastro/astro/pull/12426).

Not yet merged - milestone.

<!-- 2. Check that your PR includes `<p><Since v="4.x.0" /></p>` and imports the `<Since>` component, if necessary! -->

<!-- #### First-time contributor to Astro Docs? -->

<!-- If you are a member of the Astro Discord, please add your username in the description so we can welcome you there! -->
<!-- https://astro.build/chat -->

astrobot-houston commented:
<!-- lunaria-action-comment -->

## Lunaria Status Overview


🌕 **This pull request will trigger status changes.**

<details>
<summary>Learn more</summary>

By default, every PR changing files present in the [Lunaria configuration's `files` property](https://lunaria.dev/reference/configuration/#files-required) will be considered and trigger status changes accordingly.

You can change this by adding one of the keywords present in the [`ignoreKeywords` property in your Lunaria configuration file](https://lunaria.dev/reference/configuration/#ignorekeywords) in the PR's title (ignoring all files) or by [including a tracker directive](https://lunaria.dev/guides/tracking/#tracker-directives) in the merged commit's description. 
</details>

### Tracked Files




| File                                                                                                                                                    | Note                                                       |
| ------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------- |
| [en/guides/images.mdx](https://github.com/withastro/docs/blob/f872aecba4f13a7b18af6d6da8d0492227589926/src%2Fcontent%2Fdocs%2Fen%2Fguides%2Fimages.mdx) | Source changed, localizations will be marked as outdated.  |

<details>
<summary>Warnings reference</summary>

| Icon | Description                                                                                                                                       |
| ---- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| 🔄️  | The source for this localization has been updated since the creation of this pull request, make sure all changes in the source have been applied. |

</details>

astrobot-houston commented:
Hello! Thank you for opening your **first PR** to Astro’s Docs! 🎉

Here’s what will happen next:

1. Our GitHub bots will run to check your changes.
   If they spot any broken links you will see some error messages on this PR.
   Don’t hesitate to ask any questions if you’re not sure what these mean!

2. In a few minutes, you’ll be able to see a preview of your changes on Netlify 🥳.

3. One or more of our maintainers will take a look and may ask you to make changes.
   We try to be responsive, but don’t worry if this takes a few days.

ascorbic reviewed (commented):
This looks technically correct (which is the best kind of correct)

Princesseuh commented on src/content/docs/en/guides/images.mdx:
```diff
@@ -28,7 +28,20 @@ Store your images in the `public/` folder if you want to avoid any processing or
- For extra protection when dealing with external sources, Astro's image components and helper function will only process (e.g. optimize, transform) images from [authorized image sources specified in your configuration](#authorizing-remote-images). Remote images from other sources will be displayed with no processing.
+For extra protection when dealing with external sources, Astro's image components and helper function will only process (e.g. optimize, transform) images from [authorized image sources specified in your configuration](#authorizing-remote-images). Remote images from other sources will be displayed with no processing.
+
+#### Asset Caching
```
Might be interesting to note that local images are also cached.

sarah11918 replied:
@Princesseuh Do you think maybe an entire section on Image Caching would make sense?

Princesseuh replied:
I would say so! But it might be confusing for some users, since it's a bit more advanced than the rest and it's not something you interact with directly

Princesseuh commented:
It might be interesting to mention where the images are cached, sometimes users want to know what to need to keep between builds for caching (here you can link to `cacheDir` in the config)

oliverlynch commented:
I've moved the new documentation to its own section "Asset Caching", and mentioned how local images are also cached. 

I added a mention of `cacheDir` in the body text instead of just the readmore link -
Is it okay to write the default `cacheDir` value here, or should I avoid repeating it and have people click through to the configuration reference?

I also expanded on revalidation, explaining that the remote server requires support for the `If-Modified-Since` and `If-None-Match` headers for revalidation to work. While this is a bit technical, if someone is having difficulty getting revalidation working it is probably best for this to be clearly documented.

sarah11918 commented on src/content/docs/en/guides/images.mdx:
```diff
@@ -475,6 +475,26 @@ export default defineConfig({
 
+## Asset Caching
+
+Astro stores processed image assets in a cache directory during site builds, for both local and [remote images from authorized sources](/en/guides/images/#authorizing-remote-images). By preserving the cache directory between builds, processed assets are reused, improving build time and bandwidth usage.
+The default cache directory is `./node_modules/.astro`, however this can be changed using the `cacheDir` configuration value.
```
```suggestion
Astro stores processed image assets in a cache directory during site builds, for both local and [remote images from authorized sources](/en/guides/images/#authorizing-remote-images). By preserving the cache directory between builds, processed assets are reused, improving build time and bandwidth usage.

The default cache directory is `./node_modules/.astro`, however this can be changed using the `cacheDir` configuration value.
```

sarah11918 commented on src/content/docs/en/guides/images.mdx:
```diff
@@ -475,6 +475,26 @@ export default defineConfig({
 
+## Asset Caching
+
+Astro stores processed image assets in a cache directory during site builds, for both local and [remote images from authorized sources](/en/guides/images/#authorizing-remote-images). By preserving the cache directory between builds, processed assets are reused, improving build time and bandwidth usage.
+The default cache directory is `./node_modules/.astro`, however this can be changed using the `cacheDir` configuration value.
+
+<ReadMore>See the configuration reference for [`cacheDir`](/en/reference/configuration-reference/#cachedir)</ReadMore>
```
```suggestion
Astro stores processed image assets in a cache directory during site builds for both local and [remote images from authorized sources](/en/guides/images/#authorizing-remote-images). By preserving the cache directory between builds, processed assets are reused, improving build time and bandwidth usage.

The default cache directory is `./node_modules/.astro`, however this can be changed using the [`cacheDir`](/en/reference/configuration-reference/#cachedir) configuration setting.
```
The `cacheDir` entry isn't very exciting 😅 , so it feels like a little much to give it an entire "Read More" link!

sarah11918 commented on src/content/docs/en/guides/images.mdx:
```diff
@@ -475,6 +475,26 @@ export default defineConfig({
+
+<p><Since v="5.1.0" /></p>
+
+[Revalidation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#validation) reduces bandwidth usage and build time by checking with the remote server if an expired cached image is still up-to-date. If the server indicates that the image is still fresh, the cached version is reused, otherwise the image is redownloaded.
```
```suggestion
[Revalidation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#validation) reduces bandwidth usage and build time by checking with the remote server whether an expired cached image is still up-to-date. If the server indicates that the image is still fresh, the cached version is reused, otherwise the image is redownloaded.
```
tiny grammar nit, but in general, this is all super well written and I really appreciate it!

sarah11918 commented on src/content/docs/en/guides/images.mdx:
```diff
@@ -475,6 +475,26 @@ export default defineConfig({
+
+[Revalidation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#validation) reduces bandwidth usage and build time by checking with the remote server if an expired cached image is still up-to-date. If the server indicates that the image is still fresh, the cached version is reused, otherwise the image is redownloaded.
+
+Revalidation requires the remote server to send [Last-Modified](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Last-Modified) and/or [Etag (entity tag)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) headers with its responses, and to support the [If-Modified-Since](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since) and [If-None-Match](https:/ ...
```
```suggestion
Revalidation requires that the remote server send [Last-Modified](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Last-Modified) and/or [Etag (entity tag)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) headers with its responses. This feature is available for remote servers that support the [If-Modified-Since](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since) and [If-None-Match](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match) headers.
```
I know it looks worse in Markdown, but still tried to split this up into two more manageable chunks.

sarah11918 reviewed (commented):
Looks fantastic, @oliverlynch ! I really appreciate the great docs here, and the going above and beyond to document more than just your feature! 💜 

Had some small suggestions for your consideration, and can't wait to get this in with the next minor release! 🥳

oliverlynch commented:
Thanks @sarah11918 for the suggestions, I've committed them as-is.

sarah11918 reviewed (approved):
Fantastic @oliverlynch ! Welcome to Team Docs! :partying_face:
